[NOX to LYS] Add the NOX in-context flows to LYS

elazzabi opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds the first building blocks to support NOX in LYS. It will act as the base branch for future PRs.

### How to test the changes in this Pull Request:

No specific testing instructions for this one, each PR will have its own testing instructions.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Add the payment's new onboarding experience flow to the Launch Your Store process.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

elazzabi committed:
use endsWith instead if startsWith

This will make it easier to work with complicated URLs where the step URL is part of another URL structure

elazzabi committed:
Add source to OnboardingProvider

This can be changed with a function to get the navigation URL instead

elazzabi committed:
Pass the source to NOX in payments settings

elazzabi committed:
For the payments task, make sure to navigate to a specific URL and stop the default behaviour

elazzabi committed:
Replace hardcoded path with a * to catch everything

This will make it easier to work with LYS but also keeps backward compatibility

elazzabi committed:
Add PaymentsModal page to LYS

elazzabi committed:
Add a routes definition to open NOX when the route matches

elazzabi committed:
Merge remote-tracking branch 'origin/trunk' into add/nox-to-lys

elazzabi committed:
remove the modal in favor of sub steps

elazzabi committed:
Add includeSidebar as a top level prop for WooPaymentsOnboarding. This will be passed to Stepper

elazzabi committed:
Wrap launch your store in an OnboardingProvider

This is necessary so the content AND the sidebar can access the same context

elazzabi committed:
Add the payments-content and corresponding CSS changes/overrides

elazzabi committed:
add completed and active step icons

elazzabi committed:
add a payments sidebar that uses the context to get and display data

Also, make sure this only displays the uncompleted steps on first render

elazzabi committed:
add LYS-specific payments steps

elazzabi committed:
Change onboarding context to accept a list of steps as prop

Then change the implementation for both settings NOX and LYS NOX to use it

elazzabi committed:
Add payments content to xstate

elazzabi committed:
Add a specific case for clicking payments or woocommerce-payments, this will override everything for now, but let's adapt it later based on business logic

elazzabi committed:
Add a comment explaining the usage of some classnames

elazzabi committed:
Add xstate changes for payments sub menu

elazzabi committed:
open the payments sidebar when accessed directly via URL

There is still an edge case where steps aren't displayed for some reason

elazzabi committed:
fix dependency for Activate payments

elazzabi committed:
Fix linting issues

elazzabi committed:
Merge branch 'trunk' into add/nox-to-lys

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

elazzabi commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/data/onboarding-context.tsx:
```diff
@@ -120,11 +126,27 @@ export const OnboardingProvider: React.FC< {
-					tab: 'checkout',
-				} );
-				history.push( newPath );
+				if ( source === 'settings-payments' ) {
```
Another solution here, other than using the source, is to pass a function as a prop to take care of navigation. The function will get the `step` as an argument, and then decide the format of the new link (or even take care of the navigation itself 🤔 )

WDYT @mordeth ?

elazzabi replied:
I implemented the url strategy, looks better and more scalable https://github.com/woocommerce/woocommerce/pull/58291/commits/eeb56791ef7d91c3bfcfe7ac33c22fbfba793269

elazzabi commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/index.tsx:
```diff
@@ -36,7 +37,7 @@ export default function WooPaymentsModal( {
 	React.useEffect( () => {
 		if (
-			location.pathname.startsWith( wooPaymentsOnboardingPath ) &&
+			location.pathname.endsWith( wooPaymentsOnboardingPath ) &&
```
I'll look into how to make this more robust. We should probably compare the arg explicitly instead of the whole pathname.

elazzabi replied:
Also, the reason why I switched from `startsWith` to `endsWith` is that LYS URLs are prefixed with `launch-your-store/...`

elazzabi replied:
I changed the logic to check for the path query specifically instead of pathname https://github.com/woocommerce/woocommerce/pull/58291/commits/3905bd404b2a7df6db775f9e411fc20e2c7d3e97

elazzabi commented on plugins/woocommerce/client/admin/client/settings-payments/onboarding/providers/woopayments/steps/index.tsx:
```diff
@@ -54,3 +54,37 @@ export const steps: WooPaymentsProviderOnboardingStep[] = [
 	},
 ];
+
+export const LYSPaymentsSteps: WooPaymentsProviderOnboardingStep[] = [
```
Not sure if this is the perfect place for defining LYS steps, but it's worth keeping all step definitions in one place.

mordeth replied:
Let's move the LYS steps into the LYS folder for consistency.

elazzabi commented on plugins/woocommerce/client/admin/client/launch-your-store/hub/sidebar/tasklist.tsx:
```diff
@@ -104,8 +104,24 @@ export function taskClickedAction( event: {
 	recordEvent( 'launch_your_store_hub_task_clicked', {
 		task: event.task.id,
 	} );
+
```
Some changes in this file are only made so I can test things locally (one example is forcing the navigation to a specific URL). We have many use cases to support, so this will not be a simple URL redirection.

Should be taken care as part of #58276

elazzabi reviewed (commented):
Added a few comments to explain some parts

elazzabi committed:
remove redundant else

elazzabi committed:
Add guards to move to the payments content if navigated directly

elazzabi committed:
Use url strategy instead of source prop

elazzabi committed:
Check path directly, instead of pathname

elazzabi committed:
Move useEffect to make it easy to follow up

elazzabi committed:
Add LYS url strategy for onboarding provider

elazzabi committed:
remove non used source prop

elazzabi committed:
remove URL force refresh and pollution for a simple cleanup

elazzabi committed:
Only set the ref if we have backend steps, this way we are sure that the API call is done

elazzabi committed:
Remove cleanupPaymentsHistory as not needed

elazzabi committed:
remove commented cleanup

elazzabi committed:
add background refreshing for when moving back from the payments sub-steps

mordeth reviewed (commented):
The code changes look solid, and I believe we should use this PR’s branch as the foundation for all future improvements to NOX behavior in the LYS flow.

elazzabi committed:
[NOX in LYS] Generate a different Jetpack connection link for LYS (#58310)

* Add source to the selectors

* Pass the source to the onboarding URL if present

* Add source to query params and pass it to class functions

* Add logic to generate different URLs based on the source

* Adapt the controller to read the source and pass it down to the service

* Pass the source to the OnboardingProvider in LYS

* Pass the source to the redux store query inside the onboarding context

* Fix linting for PHP

Commit:
Add changefile(s) from automation for the following project(s): @woocommerce/data, woocommerce, woocommerce/client/admin

elazzabi committed:
[NOX in LYS] Multiple design tweaks to the NOX (#58361)

* Reduce border-radius

* remove description from sidebar

* Rename Get paid to Set up WooPayments

* Remove one extra heading in the sidebar

* remove the finish step as it's not needed for LYS

* fix order of the step

* Fix step title

* Add a placeholder component for sub-steps

* Display a placeholder if the context is loading data

* Add animation support for placeholders

* Add easing in and out for placeholders and sub-steps

* Fix linting

* Revert "Fix step title"

This reverts commit 8c32b52b5154943a1be44541c41d2aee854f5c47.

* Reapply "Fix step title"

This reverts commit 0021468ae3e0226951fe39652bbedd38425cebfe.

* Fix paddings for the placeholder to mimic those of the steps

* Update animation to follow design updates

* Remove extra logic to display only uncompleted steps

* Update completed step styling and task icon

mordeth committed:
[NOX in LYS] Introduce installation step for WooPayments if the plugin is not active (#58379)

* Introduce installation step on LYS for WooPayments if the plugin is not active

elazzabi committed:
Display finished steps first

vladolaru committed:
Merge branch 'trunk' into add/nox-to-lys

mordeth committed:
[NOX in LYS] Redirect to the LYS hub after successful account onboarding (#58496)

* Redirect to LYS hub after successful onboarding

elazzabi committed:
[NOX in LYS] Add cases when to show NOX in LYS (#58438)

* Send SHOW_PAYMENTS only if WooPayments is enabled

* Remove task clicked override for a simple execution and update the task clicked function to transition to SHOW_PAYMENTS when necessary

* remove unused import

* Also display the NOX if the country is a WCPay supported one

* Move the rendering logic to the backend

* Add has_gateways_other_than_woopayments function

* Potentially mark this as complete only if it has a gateway other than woopayments

We'll need to add more checks afterward

* Bump time from 2 minutes to 5

* add is_woopayments_active function

* Add is_store_in_woopayments_supported_geo function

* Add necessary imports

* Add is_woopayments_test_account function

* Add is_woopayments_configured function

* Add logic for the action_url

The '' will push the frontend to do the navigation. While the link will kick a redirection

* use str_starts_with to look for woopayments gateways

* remove unused function

* Fix linting issues

* Continue linting fixes

* remove todo to fix linting

* Fix function signature

* remoce unnecessary comment

* Make check more robust for WooPayments class call

* Add get_woopayments_gateway_ids helper function

* Add do_other_gateways_need_setup helper function

* Implement the last use case

* Fix linting issues

* refac: Change the Payments task LYS approach to let the JS control the LYS behavior

* implement display logic on the frontend

* add a check for woopayments not active

* Add a top check for the supported country

* If woopayments has a test account, do not mark it as complete

* Rename function to has_online_gateways

* refac: Return early

* Add has_gateways back

* Merge branch 'add/use-cases-for-nox-lys' of https://github.com/woocommerce/woocommerce into add/use-cases-for-nox-lys

* revert the name change

* Clean up unused method

* Add WooPayments onboarding flag for transitioning from test to live

* Add provider onboarding flags

* Add payments task method to determine if WooPayments is transitioning from test to live

* Lint fixes

* Revert "revert the name change"

This reverts commit 453c6d4d055e4fa92446bb1772b75e6449a94281.

* Revert "Lint fixes"

This reverts commit b5b2de1fe76538fb4836e7a8ceb1e90191e35967.

* Revert "Add payments task method to determine if WooPayments is transitioning from test to live"

This reverts commit 453a1f5b9b211300013896c85c3192ae647c1d9a.

* Revert "Add provider onboarding flags"

This reverts commit ac4a681a821a5d9af0e4dc6eb54d12a68d0b37f0.

* Revert "Add WooPayments onboarding flag for transitioning from test to live"

This reverts commit fe3b2f11de4c05cec88a8f2e3fc173fc6c492d3c.

* Override payments task status and title from the frontend for LYS

* Remove non-existent task

* Add Payments task logic to determine if WooPayments is installed

* Add support for enable/install/installing

* Move function up

---------

Co-authored-by: Vlad Olaru <vlad@pixelgrade.com>

elazzabi committed:
Merge branch 'trunk' into add/nox-to-lys

mordeth reviewed (approved):
:shipit:

P.S. I'm getting the Get paid task completed after onboarding a WooPayments test account, is that expected?
